### ISSUE TITLE
IDA integration and core dumps/absolute addresses

### ISSUE THREAD

**Issue by nikvoid** [pwndbg/pwndbg#1688]
I want to point out to several issues, that one can experience trying to use IDA integration when exploring core dumps (or images rebased in IDA to match virtual memory address);
1. ~~Symbol names are fetched only when user makes instruction step (I didn't make deep dive into pwndbg code, this is just observations) . When exploring core dumps, IDA integration becomes partially useless because user cannot make step to next instruction and trigger symbol fetch.~~
Only symbols of main executable are being fetched
(fixed in diff)
2. When image in IDA rebased to match core dump adresses, it seems like `l2r` and `r2l` address fixers corrupt address. Maybe add a config value to disable address fixups? 
(fixed in diff)
(may be continued)

**Comment by disconnect3d** [comment]
Hi, the IDA integration wasn't touched for a long time, mostly because I don't use IDA with Python scripting since it is not available in the free version that can be downloaded.

Anyway, I would accept PRs if anyone wants to improve the state of Pwndbg+IDA integration. 

Also, perhaps we should just remove IDA integration from Pwndbg and rely on https://github.com/mahaloz/decomp2dbg? Or integrate this project into Pwndbg somehow?

@nikvoid If you are using IDA, can you check decomp2dbg and investigate if we can integrate it here?

**Comment by nikvoid** [comment]
That seems interesting, thanks for pointing out to it. I'll try to use decomp2dbg for now instead.
> @nikvoid If you are using IDA, can you check decomp2dbg and investigate if we can integrate it here?

I will report my experience. 
My current use case is investigating some nasty obfuscated android native lib, that loads several independent code sections dynamically.
So I decided to add those sections to IDA alongside the main binary and then fix up addresses.
The main problem is that I either need to keep GDB session alive all time or need to fix up all addresses somehow each time I start new session. And to avoid these issues I dumped memory when program was in state I'm interested in and now I am trying to use pwndbg + IDA

**Comment by nikvoid** [comment]
Well, I tried.
I have to say: I like your way more.
I had a bunch of problems using decomp2dbg. Few of them was the same ones that I had with the pwndbg-IDA, namely address rebasing (now I just disabled it, to be able to continue). Also there was a problem with passing around 64-bit ints with XMLRPC (for both pwndbg and D2D I used one workaround, more on that below). And finally, I needed to wait about 30 minutes for D2D server to import all data from IDA. Each time I start it. I don't know whether this is IDA API problem or D2D's, but I think it's unuseable for large IDB's. 

But I have some positive outcomes from this situation: I managed to fix pwndbg for my usecase. Maybe I'll land a PR someday, 
now I will just share changes that I made, for other people with the same problem. 

About XMLRPC: I'm encoding addresses to strings with hex representation, to avoid sending them as ints and to fit too large addresses into XMLRPC constraints.

P.S. I only managed to get D2D decompile view working. Displaying variable names from decompile result seems like interesting feature, but I failed to get it working along with symbol names, currently. Also note that I'm hosting GDB on android phone, so D2D becomes even slower.

**Comment by nikvoid** [comment]
diffs:
```diff
diff --git a/ida_script.py b/ida_script.py
index e8b4330..a8f0d0b 100644
--- a/ida_script.py
+++ b/ida_script.py
@@ -49,14 +49,21 @@ xmlclient.Marshaller.dispatch[type(1 << 63)] = create_marshaller("<value><i8>%d<
 xmlclient.Marshaller.dispatch[type(0)] = create_marshaller("<value><i8>%d</i8></value>")
 xmlclient.Marshaller.dispatch[idaapi.cfuncptr_t] = create_marshaller(just_to_str=True)
 
-host = "127.0.0.1"
+host = "0.0.0.0"
 port = 31337
 
 mutex = threading.Condition()
 
-
 def wrap(f):
+    # convert hex to int
+    def conv_addr(a):
+        if type(a) == str and a.startswith("0x"):
+            return int(a, 16)
+        else:
+            return a
+
     def wrapper(*a, **kw):
+        a = [conv_addr(arg) for arg in a]
         rv = []
         error = []
 
diff --git a/pwndbg/gdblib/symbol.py b/pwndbg/gdblib/symbol.py
index 572abe4..1995429 100644
--- a/pwndbg/gdblib/symbol.py
+++ b/pwndbg/gdblib/symbol.py
@@ -77,12 +77,10 @@ def get(address: int, gdb_only=False) -> str:
 
     if not gdb_only and result.startswith("No symbol"):
         address = int(address)
-        exe = pwndbg.gdblib.elf.exe()
-        if exe:
-            exe_map = pwndbg.gdblib.vmmap.find(exe.address)
-            if exe_map and address in exe_map:
-                res = pwndbg.ida.Name(address) or pwndbg.ida.GetFuncOffset(address)
-                return res or ""
+        map = pwndbg.gdblib.vmmap.find(address)
+        if map and address in map:
+            res = pwndbg.ida.Name(address) or pwndbg.ida.GetFuncOffset(address)
+            return res or ""
 
     # If there are newlines, which means that there are multiple symbols for the address
     # then use the first one (see also #1610)
diff --git a/pwndbg/ida.py b/pwndbg/ida.py
index 270d87e..b90a205 100644
--- a/pwndbg/ida.py
+++ b/pwndbg/ida.py
@@ -38,10 +38,6 @@ ida_timeout = pwndbg.gdblib.config.add_param(
     "ida-timeout", 2, "time to wait for ida xmlrpc in seconds"
 )
 
-xmlrpc.client.Marshaller.dispatch[int] = lambda _, v, w: w("<value><i8>%d</i8></value>" % v)
-
-xmlrpc.client.Marshaller.dispatch[type(0)] = lambda _, v, w: w("<value><i8>%d</i8></value>" % v)
-
 _ida = None
 
 # to avoid printing the same exception multiple times, we store the last exception here
@@ -177,19 +173,12 @@ def can_connect() -> bool:
 
 
 def l2r(addr):
-    exe = pwndbg.gdblib.elf.exe()
-    if not exe:
-        raise Exception("Can't find EXE base")
-    result = (addr - int(exe.address) + base()) & pwndbg.gdblib.arch.ptrmask
-    return result
-
+    # do not rebase, convert to hex
+    return hex(addr)
 
 def r2l(addr):
-    exe = pwndbg.gdblib.elf.exe()
-    if not exe:
-        raise Exception("Can't find EXE base")
-    result = (addr - base() + int(exe.address)) & pwndbg.gdblib.arch.ptrmask
-    return result
+    # do not rebase, convert back to int
+    return int(addr, 16)
 
 
 def remote(function) -> None:
```

**Comment by disconnect3d** [comment]
TL;DR: Closing all IDA integration issues for now. We implemented Binja integration recently and an abstraction for `IntegrationProvider`. I would be happy of accepting a new version of IDA sync if anyone is interested in implementing it (since the old script and code most likely do not work at all).

https://github.com/pwndbg/pwndbg/issues/218#issuecomment-2518778108
